scripts/create_x509_chain_crypto.py

```diff
@@ -0,0 +1,558 @@
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import rsa
+from cryptography.x509.oid import NameOID
+from datetime import datetime, timedelta, time
+import uuid
+import os
+import argparse
+import getpass
+import base64
+import shutil
+
+"""
+For using this script must have cryptography package installed in python environment.
+"""
+
+PUBLIC_EXPONENT = 65537
+EXTENSION_NAME = ".pem"
+COMMON_DEVICE_PASSWORD_FILE = "demoCA/private/device_key"
+COMMON_DEVICE_CSR_FILE = "demoCA/newcerts/device_csr"
+COMMON_DEVICE_CERT_FILE = "demoCA/newcerts/device_cert"
+
+
+def create_private_key(password_file, password=None, key_size=4096):
+    if password:
+        encrypt_algo = serialization.BestAvailableEncryption(str.encode(password))
+    else:
+        encrypt_algo = serialization.NoEncryption()
+
+    private_key = rsa.generate_private_key(
+        public_exponent=PUBLIC_EXPONENT, key_size=key_size, backend=default_backend()
+    )
+    # Write our key to file
+    with open(password_file, "wb") as f:
+        f.write(
+            private_key.private_bytes(
+                encoding=serialization.Encoding.PEM,
+                format=serialization.PrivateFormat.TraditionalOpenSSL,
+                encryption_algorithm=encrypt_algo,
+            )
+        )
+
+    return private_key
+
+
+def create_certificate_chain(
+    common_name,
+    ca_password,
+    intermediate_password,
+    device_password,
+    device_count=1,
+    key_size=4096,
+    days=3650,
+):
+    """
+    This method will create a basic 3 layered chain certificate containing a root, then an intermediate and then some number of leaf certificates.
+    This function is only used when the certificates are created from script.
+
+    :param common_name: The common name to be used in the subject. This is a single common name which would be applied to all certs created. Since this common name is meant for all,
+    this common name will be prepended by the words "root", "inter" and "device" for root, intermediate and device certificates.
+    For device certificates the common name will be further appended with the index of the device.
+    :param ca_password: The password for the root certificate which is going to be referenced by the intermediate.
+    :param intermediate_password: The password for the intermediate certificate
+    :param device_password: The password for the device certificate
+    :param device_count: The number of leaf devices for which that many number of certificates will be generated.
+    :param key_size: The key size to use for encryption. The default is 4096.
+    :param days: The number of days for which the certificate is valid. The default is 1 year or 365 days.
+    For the root cert this value is multiplied by 10. For the device certificates this number will be divided by 10.
```

Comment:
>For the root cert this value is multiplied by 10. For the device certificates this number will be divided by 10. [](start = 4, length = 112)

why?  i know this is for testing only, but I'm not sure I follow the logic.

Reply:
now removed and everything is default to 1 year...i think i took a suggestion of creating root for 10 years..not a good idea.....

---
In reply to: [432499532](https://github.com/Azure/azure-iot-sdk-python/pull/574#discussion_r432499532) [](ancestors = 432499532)